include/flatcc/flatcc_assert.h

```diff
@@ -0,0 +1,39 @@
+#ifndef FLATCC_ASSERT_H
+#define FLATCC_ASSERT_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+* These assert abstraction is __only__ for runtime libraries.
+*
+* The flatcc compiler uses Posix assert routines regardless
+* of how this file is configured.
+*
+* This header makes it possible to use systems where assert is not
+* valid to use.
+*
+* `FLATCC_ASSERT` is designed to handle errors which cannot be ignored
+* and could lead to crash.
+*
+* `FLATCC_ASSERT` can be overrided by compiler definition as well as
+* `FLATCC_ASSERT_INCLUDE` to include needed headers with own assert handler
+* declaration
+*
+* If `FLATCC_ASSERT` definition is not providede, posix assert is being used
+*/
+
+#if !defined(FLATCC_ASSERT)
+#include <assert.h>
+#define FLATCC_ASSERT assert
+#elif defined(FLATCC_ASSERT_INCLUDE)
+     // Include file with forward declaration
```

Comment:
Please  use C style comments

Reply:
I'm not convinced that defining an include file is a good idea. It is not being used anywhere else, such as custom allocation. It's also possible to override https://github.com/dvidelabs/flatcc/blob/master/config/config.h

Reply:
Isn't that config.h used for compiler setup? Include file is needed for runtime part as custom assert can be overridden by some already declared function in existing header.

Reply:
Yes, I think you are right. Maybe there is something missing. I'm still not comfortable with controlling the build environment via defines in this manner, especially when it hasn't been necessary until know. You should be able to add your own files to build and use defines to take advantage of that implementation, which is why #ifndef is widely used. I'm open to suggestions.

Reply:
Please have a look at https://github.com/dvidelabs/flatcc/blob/master/include/flatcc/flatcc_rtconfig.h

Reply:
I think it can be done as it is. flatcc_alloc.h also has ifdefs for each functions to provide default implementation. In case of default include file it is strictly associated with default assert implementation which comes from posix. There is no point to have FLATCC_ASSERT_INCLUDE  defined in that case. If you would like to keep FLATCC_ASSERT_INCLUDE it would require to keep also an additional define flag to enable its usage.  flatcc_rtconfig.h seems to be not used in builder, or did I miss something?

Reply:
From memory, flatcc_rtconfig.h is only used for runtime, however, the compiler uses libflatcc which does include runtime components of libflatccrt, so indirectly it is used everywhere.

Reply:
If you are happy to remove the FLATCC_ASSERT_INCLUDE section, I think this PR is close to OK, and I can give it a closer look.

Reply:
I need to inject somehow a header include in each place where custom assert is being used. Otherwise I'll end up with using undefined symbol. I see no other way to do that than by that define. In the same time, I do not want to add that include if I want to use posix assert. Adding #include only in case when FLATCC_ASSERT_INCLUDE is defined seems to be the only option to solve that.

Reply:
How about 
```
#ifdef FLATCC_CUSTOM_CONFIG
#include "flatcc_custom_config"
#endif

In flatcc/flatcc_rtconfig.h?

You can then add an include path and a custom header. If you do this, please check that it works with alt. assert.

Reply:
Sounds like lot of steps just to include one header. It would be hard to track all dependencies of headers inclusion.

Reply:
But it will work with allocation etc. as well.

Reply:
And, it is actually just two CFLAGS: -Imypath -DFLATCC_CUSTOM_RTCONFIG
You also change include path and have a verifier traces enabled etc.

Reply:
Maybe a build option is also needed for custom include path.

Reply:
> Maybe a build option is also needed for custom include path.

which can be handled directly in flatcc_assert.h :) if you would like to have defined it in header, you can always add that define inflatcc/flatcc_rtconfig.h

I would add separate preprocessor option for alloc include path as usually memory related functions will be defined in other header than assert. 

Where in current implementation is desired place to define override alloc functions from flatcc_alloc.h? You have it ifdeffed just like I want to have that include for custom assert header, but I don't now where would you place that definitions if not as compile option.

Reply:
there are endless such includes - alloc is just the most obvious example - that is why I dislike the approach.

Reply:
In README.md there is a statement: 

> while piloting a new platform
> flatcc_alloc.h can be used to override runtime allocation functions

. So I guess it is not intended to configure alloc in flatcc_rtconfig.h. I think it can also be done this way for assert. Flatcc_rtconfig.h seems to be used only in parser and verifier. I do not feel that confident to get  through all headers and dependencies and make it also working for builder and reader.

My only proposal is to keep that FLATCC_ASSERT_INCLUDE to have it aligned with Google's flatbuffers approach. If you are not interested, I can remove it and I'll add it in my local implementation.

Reply:
Yes please remove. I think it is better for you to privately customise flatcc_assert.h in this case - which is also one reason to separate out the logic. I would like to make this simpler, but as I mentioned, adding special handling in build system for every imaginable configuration is not viable.